src/geode/mesh/helpers/euclidean_distance_transform.cpp

```diff
@@ -0,0 +1,406 @@
+    {
+        using Index = GridCellIndices< dimension >;
+
+    public:
+        EuclideanDistanceTransform( const RegularGrid< dimension >& grid,
+            absl::Span< const GridCellIndices< dimension > > grid_cell_id,
+            absl::string_view distance_map_name )
+            : grid_{ grid }, squared_cell_length_{}, distance_map_{ nullptr }
+        {
+            for( const auto d : Range( dimension ) )
+            {
+                squared_cell_length_[d] = grid_.cell_length_in_direction( d )
+                                          * grid_.cell_length_in_direction( d );
+            }
+            distance_map_ =
+                grid_.cell_attribute_manager()
+                    .template find_or_create_attribute< VariableAttribute,
+                        double >( distance_map_name,
+                        std::numeric_limits< double >::max() );
+            for( const auto& cell_id : grid_cell_id )
+            {
+                distance_map_->set_value( grid_.cell_index( cell_id ), 0. );
+            }
+        }
+
+        std::shared_ptr< VariableAttribute< double > > distance_map() const
+        {
+            return distance_map_;
+        }
+
+        void compute_squared_distance_map();
+
+        void squared_root_filter()
+        {
+            async::parallel_for(
+                async::irange( index_t{ 0 }, grid_.nb_cells() ),
+                [this]( index_t cell ) {
+                    distance_map_->modify_value( cell, []( double& value ) {
+                        value = std::sqrt( value );
+                    } );
+                } );
+            return;
+        }
+
+    private:
+        void propagate_directional_squared_distance( const index_t direction );
+
+        double propagate_directional_step_squared_distance(
+            const Index& from_index,
+            const Index& to_index,
+            const index_t direction,
+            const double last_step_squared_distance )
+        {
+            const auto old_distance =
+                distance_map_->value( grid_.cell_index( from_index ) );
+            const auto step_squared_distance =
+                old_distance == 0 ? squared_cell_length_[direction]
+                                  : last_step_squared_distance
+                                        + 2 * squared_cell_length_[direction];
+            const auto new_distance = old_distance + step_squared_distance;
+            distance_map_->modify_value(
+                grid_.cell_index( to_index ), [new_distance]( double& value ) {
+                    value = std::min( value, new_distance );
+                } );
+            return step_squared_distance;
+        }
+
+        void combine_squared_distance_components( const index_t direction );
+
+        double directional_squared_distance( const index_t from,
+            const index_t to,
+            const index_t direction ) const
+        {
+            return ( static_cast< double >( from )
```

Comment:
maybe cast once from and to and then use it twice.

Reply:
ok